src/Amalgam/evaluablenode/EvaluableNodeManagement.cpp

```diff
@@ -893,59 +896,95 @@ std::pair<bool, bool> EvaluableNodeManager::UpdateFlagsForNodeTreeRecurse(Evalua
 	}
 }
 
-void EvaluableNodeManager::MarkAllReferencedNodesInUseRecurse(EvaluableNode *tree)
+void EvaluableNodeManager::MarkAllReferencedNodesInUse(EvaluableNode *tree)
 {
-#ifdef AMALGAM_FAST_MEMORY_INTEGRITY
-	assert(tree->IsNodeValid());
-#endif
+	if(tree == nullptr)
+		return;
 
-	//if entering this function, then the node hasn't been marked yet
 	tree->SetKnownToBeInUse(true);
+	auto &node_stack = threadLocalAllocationBuffer;
+	node_stack.push_back(tree);
 
-	auto type = tree->GetType();
-	if(DoesEvaluableNodeTypeUseOrderedData(type))
+	while(!node_stack.empty())
 	{
-		for(auto &e : tree->GetOrderedChildNodesReference())
+		auto *node = node_stack.back();
+	#ifdef AMALGAM_FAST_MEMORY_INTEGRITY
+		assert(node->IsNodeValid());
+	#endif
+		node_stack.pop_back();
+
+		auto type = node->GetType();
+		if(DoesEvaluableNodeTypeUseOrderedData(type))
 		{
-			if(e != nullptr && !e->GetKnownToBeInUse())
-				MarkAllReferencedNodesInUseRecurse(e);
+			for(auto &cn : node->GetOrderedChildNodesReference())
+			{
+				if(cn != nullptr && !cn->GetKnownToBeInUse())
+				{
+					cn->SetKnownToBeInUse(true);
+					node_stack.push_back(cn);
+				}
+			}
 		}
-	}
-	else if(DoesEvaluableNodeTypeUseAssocData(type))
-	{
-		for(auto &[_, e] : tree->GetMappedChildNodesReference())
+		else if(DoesEvaluableNodeTypeUseAssocData(type))
 		{
-			if(e != nullptr && !e->GetKnownToBeInUse())
-				MarkAllReferencedNodesInUseRecurse(e);
+			for(auto &[_, cn] : node->GetMappedChildNodesReference())
+			{
+				if(cn != nullptr && !cn->GetKnownToBeInUse())
+				{
+					cn->SetKnownToBeInUse(true);
+					node_stack.push_back(cn);
+				}
+			}
 		}
 	}
 }
 
 #ifdef MULTITHREAD_SUPPORT
-void EvaluableNodeManager::MarkAllReferencedNodesInUseRecurseConcurrent(EvaluableNode *tree)
+void EvaluableNodeManager::MarkAllReferencedNodesInUseConcurrent(EvaluableNode *tree)
 {
+	if(tree == nullptr)
+		return;
+
 #ifdef AMALGAM_FAST_MEMORY_INTEGRITY
 	assert(tree->IsNodeValid());
 #endif
 
-	//if entering this function, then the node hasn't been marked yet
-	tree->SetKnownToBeInUseAtomic(true);
+	tree->SetKnownToBeInUse(true);
+	//because marking occurs during garbage collection and threadLocalAllocationBuffer
```

Comment:
I think we should add a comment by the declaration of the threadLocalAllocationBuffer itself that it is also being used for this purpose. If some one makes changes to how it works in the future, its going to be pretty easy to overlook this.